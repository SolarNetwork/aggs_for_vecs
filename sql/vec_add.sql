-- int add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);

-- float add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);

-- numeric add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
