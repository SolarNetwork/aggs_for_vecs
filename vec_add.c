
Datum vec_add(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_add);

/**
 * Add elements of two arrays.
 *
 * This function takes two arrays of n-elements and returns an array of
 * n-elements where each element is the sum of the input elements
 * of the same index.
 */
Datum
vec_add(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int valsLength;
  ArrayType *leftArray, *rightArray, *retArray;
  Datum *leftContent, *rightContent, *retContent;
  bool *leftNulls, *rightNulls, *retNulls;
  int i;
  int dims[1];
  int lbs[1];

  if (PG_ARGISNULL(0) || PG_ARGISNULL(1)) {
    PG_RETURN_NULL();
  }

  leftArray = PG_GETARG_ARRAYTYPE_P(0);
  rightArray = PG_GETARG_ARRAYTYPE_P(1);

  if (ARR_NDIM(leftArray) == 0 || ARR_NDIM(rightArray) == 0) {
    PG_RETURN_NULL();
  }
  if (ARR_NDIM(leftArray) > 1 || ARR_NDIM(rightArray) > 1) {
    ereport(ERROR, (errmsg("vec_add: one-dimensional arrays are required")));
  }

  elemTypeId = ARR_ELEMTYPE(leftArray);

  if (elemTypeId != INT2OID &&
      elemTypeId != INT4OID &&
      elemTypeId != INT8OID &&
      elemTypeId != FLOAT4OID &&
      elemTypeId != FLOAT8OID &&
      elemTypeId != NUMERICOID) {
    ereport(ERROR, (errmsg("vec_add input must be array of SMALLINT, INTEGER, BIGINT, REAL, DOUBLE PRECISION, or NUMERIC")));
  }
  if (rightArray && elemTypeId != ARR_ELEMTYPE(rightArray)) {
    ereport(ERROR, (errmsg("vec_add mins array must be the same type as input array")));
  }

  valsLength = (ARR_DIMS(leftArray))[0];
  if (rightArray && valsLength != (ARR_DIMS(rightArray))[0]) {
    ereport(ERROR, (errmsg("vec_add right array must be the same length as left array")));
  }

  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);

  deconstruct_array(leftArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &leftContent, &leftNulls, &valsLength);
  if (rightArray) {
    deconstruct_array(rightArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
        &rightContent, &rightNulls, &valsLength);
  }

  retContent = palloc0(sizeof(Datum) * valsLength);
  retNulls = palloc0(sizeof(bool) * valsLength);

  for (i = 0; i < valsLength; i++) {
    if (leftNulls[i] || rightNulls[i]) {
      retNulls[i] = true;
      continue;
    }
    retNulls[i] = false;
    switch(elemTypeId) {
      case INT2OID:
        retContent[i] = Int16GetDatum(DatumGetInt16(leftContent[i]) + DatumGetInt16(rightContent[i]));
        break;
      case INT4OID:
        retContent[i] = Int32GetDatum(DatumGetInt32(leftContent[i]) + DatumGetInt32(rightContent[i]));
        break;
      case INT8OID:
        retContent[i] = Int64GetDatum(DatumGetInt64(leftContent[i]) + DatumGetInt64(rightContent[i]));
        break;
      case FLOAT4OID:
        retContent[i] = Float4GetDatum(DatumGetFloat4(leftContent[i]) + DatumGetFloat4(rightContent[i]));
        break;
      case FLOAT8OID:
        retContent[i] = Float8GetDatum(DatumGetFloat8(leftContent[i]) + DatumGetFloat8(rightContent[i]));
        break;
      case NUMERICOID:
#if PG_VERSION_NUM < 120000
        retContent[i] = DirectFunctionCall2(numeric_add, leftContent[i], rightContent[i]);
#else
        retContent[i] = NumericGetDatum(numeric_add_opt_error(DatumGetNumeric(leftContent[i]), DatumGetNumeric(rightContent[i]), NULL));
#endif
        break;
    }
  }

  dims[0] = valsLength;
  lbs[0] = 1;

  retArray = construct_md_array(retContent, retNulls, 1, dims, lbs,
      elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode);

  PG_RETURN_ARRAYTYPE_P(retArray);
}

