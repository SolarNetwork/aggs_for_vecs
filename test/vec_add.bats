load test_helper

@test "int16 add" {
  result="$(query "SELECT vec_add(ARRAY[1,3,5]::smallint[], ARRAY[2,4,6]::smallint[])")";
  echo $result;
  [ "$result" = "{3,7,11}" ]
}

@test "int32 add" {
  result="$(query "SELECT vec_add(ARRAY[1,3,5]::integer[], ARRAY[2,4,6]::integer[])")";
  echo $result;
  [ "$result" = "{3,7,11}" ]
}

@test "int64 add" {
  result="$(query "SELECT vec_add(ARRAY[1,3,5]::bigint[], ARRAY[2,4,6]::bigint[])")";
  echo $result;
  [ "$result" = "{3,7,11}" ]
}

@test "float32 add" {
  result="$(query "SELECT vec_add(ARRAY[1.0,2.4,3.4]::real[], ARRAY[3.4,0.5,0.9]::real[])")";
  echo $result;
  [ "$result" = "{4.4,2.9,4.3}" ]
}

@test "float64 add" {
  result="$(query "SELECT vec_add(ARRAY[1.1,2.2,3.4]::float[], ARRAY[3.4,2.2,1.1]::float[])")";
  echo $result;
  [ "$result" = "{4.5,4.4,4.5}" ]
}

@test "numeric add" {
  result="$(query "SELECT vec_add(ARRAY[1.23,2.234,3.456]::numeric[], ARRAY[3.456,2.234,1.123]::numeric[])")";
  echo $result;
  [ "$result" = "{4.686,4.468,4.579}" ]
}

@test "numeric add measurements" {
  result="$(query "SELECT tot FROM (SELECT vec_add(nums, lag(nums) OVER (ORDER BY sensor_id)) AS tot FROM measurements d WHERE sensor_id IN (3,4)) d WHERE d.tot IS NOT NULL")";
  echo $result;
  [ "$result" = "{2.46,NULL,5.79}" ]
}
