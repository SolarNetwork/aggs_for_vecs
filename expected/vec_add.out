-- int add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      | pg_typeof 
------------+-----------
 {6,NULL,7} | integer[]
(1 row)

-- int add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      | pg_typeof 
------------+-----------
 {6,NULL,7} | integer[]
(1 row)

-- int add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      | pg_typeof 
------------+-----------
 {2,NULL,4} | integer[]
(1 row)

-- int add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | integer[]
(1 row)

-- int add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | integer[]
(1 row)

-- int add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(ints, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], ints) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::int[], NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- float add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      |     pg_typeof      
------------+--------------------
 {6,NULL,7} | double precision[]
(1 row)

-- float add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      |     pg_typeof      
------------+--------------------
 {6,NULL,7} | double precision[]
(1 row)

-- float add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      |     pg_typeof      
------------+--------------------
 {2,NULL,4} | double precision[]
(1 row)

-- float add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         |     pg_typeof      
------------------+--------------------
 {NULL,NULL,NULL} | double precision[]
(1 row)

-- float add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         |     pg_typeof      
------------------+--------------------
 {NULL,NULL,NULL} | double precision[]
(1 row)

-- float add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(floats, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], floats) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::float[], NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- numeric add to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {6.23,NULL,7.34} | numeric[]
(1 row)

-- numeric add to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {6.23,NULL,7.34} | numeric[]
(1 row)

-- numeric add to array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {2.46,NULL,4.68} | numeric[]
(1 row)

-- numeric add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | numeric[]
(1 row)

-- numeric add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | numeric[]
(1 row)

-- numeric add scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(5::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric add scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric add to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(nums, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric add to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], nums) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric add to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric add to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric add to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_add(NULL::numeric[], NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

